gh-1838 added checks for pipe streams

llvdl opened the pull request:
Fixes #1838

ddalek commented:
Since Slim's implementation has methods beyond what's in StreamInterface anyway, I'd add isSeekableForward() and seekForward() for completeness. If one's dealing with pipes, he/she might need them anyway.

```
    /**
     * Returns whether or not the stream is seekable forward.
     *
     * Note: This method is not part of the PSR-7 standard
     * and is provided for Stream's compatibility with pipes.
     *
     * @return bool
     */
    public function isSeekableForward()
    {
        return $this->isSeekable() || $this->isPipe();
    }

    /**
     * Seek forward in the stream.
     *
     * Note: This method is not part of the PSR-7 standard
     * and is provided for Stream's compatibility with pipes.
     *
     * @param int $offset Stream offset
     * @throws RuntimeException on failure.
     */
    public function seekForward($offset)
    {
        // Note that fseek returns 0 on success!
        if (!$this->isSeekableForward() || fseek($this->stream, $offset, SEEK_CUR) === -1) {
            throw new RuntimeException('Could not seek forward in stream');
        }
    }
```

I'd also change isSeekable() implementation - there's no need for isPipe() call, I think, as 'seekable' metadata is guaranteed to be 1 in case of seekable handle and something else (typically 0 but undef for pipes) in other cases.

```
    public function isSeekable()
    {
        if ($this->seekable === null) {
            $this->seekable = false;
            if ($this->isAttached()) {
                $meta = $this->getMetadata();
                $this->seekable = ($meta['seekable'] === 1);
            }
        }

        return $this->seekable;
    }
```

HTH.

llvdl commented:
Thanks for the feedback, @ddalek.

I replaced `isSeekable()` with your version, but then multiple unit tests fail, so I reverted to the version I submitted before.

ddalek commented:
Argh, I misinterpreted print_r() output. My comparison was unnecessary, the value returned is bool, not int. This will teach me not to rely on print_r(). :)

Still, 'seekable' meta should be false for pipes (despite SEEK_CUR working) and `&& !$this->isPipe()` in isSeekable() adds the cost of fstat() to the first isSeekable() call on verifiably seekable streams. But that's just a tiny perf nitpick (if I'm not talking out of my back again).

llvdl commented:
For some reason the tests fail on the hhvm machine: `RuntimeException: Could not seek forward in stream` is raised when trying to read from the pipe.

llvdl commented:
I removed the check for `&& !$this->isPipe()` and the test still pass, so it seems you are right, @ddalek .

ddalek commented:
Cursory search through their issues shows quite a few historical incompatibilities around files and streams, including a new one around fseek() so I'm not terribly surprised. I'll take the time over the weekend to see if there's anything that can and should be done in Slim or if it's HHVM problem.

akrabat commented:
Tests fail on HHVM.

llvdl commented:
I created a virtual machine with hhvm and could reproduce the failing tests. The cause was `isReadable()` returning `false` in hhvm. I added a unit test and a check for `isPipe()` and the tests 

I removed the seekForward implementation for now, as it doesn't seem to work for pipes in PHP.

akrabat commented on Slim/Http/Stream.php:
```diff
@@ -73,6 +78,13 @@ class Stream implements StreamInterface
+     *
+     * @var type
+     */
+    protected $pipe;
```
Please name this property `isPipe`.

llvdl replied:
That would collide with the `isPipe` method

akrabat replied:
No. PHP can tell the difference between a property and a method with the same name…

llvdl replied:
I didn't know that :)

akrabat commented on Slim/Http/Stream.php:
```diff
@@ -20,6 +20,11 @@
 class Stream implements StreamInterface
 {
     /**
+     * This is octal as per header stat.h
```
Can you add info on what this particular mode means?

llvdl replied:
I expanded the comment

akrabat commented on Slim/Http/Stream.php:
```diff
@@ -73,6 +78,13 @@ class Stream implements StreamInterface
     /**
+     * Is this stream a pipe?
+     *
+     * @var type
```
This is a `boolean` I think

akrabat commented on Slim/Http/Stream.php:
```diff
@@ -158,6 +170,7 @@ public function detach()
         $this->writable = null;
         $this->seekable = null;
         $this->size = null;
+        $this->pipe = null;
```
Should default to `false`.

llvdl replied:
The value is cached. If `null` the value is determined and then set in `isPipe()`.

akrabat replied:
Ah. In that case. When is it cleared?

llvdl replied:
In `Stream::detach()`
